File: sdq-api-rust/db/src/goal_service.rs
use async_trait::async_trait;
use sdq_model::{
    Assessor, DemographicFilter, GboSubmission, Goal, GoalProgress, GoalType, SdqError,
};
use sdq_service::goal::GoalService;
use sqlx::PgPool;
use uuid::Uuid;

pub struct GoalServiceSqlxImpl {
    pub pool: PgPool,
}

impl GoalServiceSqlxImpl {
    pub fn new(pool: PgPool) -> GoalServiceSqlxImpl {
        GoalServiceSqlxImpl { pool }
    }
}

#[async_trait]
impl GoalService for GoalServiceSqlxImpl {
    async fn save_goal(&self, _goal: Goal) -> Result<(), SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_for_client(&self, _client_id: &Uuid) -> Result<Vec<Goal>, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn delete_all_goals(&self) -> Result<(), SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_goals_with_progress(
        &self,
        _assessors: &Vec<Assessor>,
        _filters: &Vec<DemographicFilter>,
        _min_progress: u64,
        _goal_types: &Vec<GoalType>,
        _from: &chrono::NaiveDate,
        _to: &chrono::NaiveDate,
    ) -> Result<Vec<GoalProgress>, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_goals_with_progress_for_client(
        &self,
        _client_id: &Uuid,
        _assessors: Assessor,
    ) -> Result<Vec<GoalProgress>, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_goal_progress(
        &self,
        _goal_id: &Uuid,
        _assessor: Assessor,
    ) -> Result<GoalProgress, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn update_goal(&self, _goal: Goal) -> Result<Goal, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_goal(&self, _goal_id: &Uuid) -> Result<Goal, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn submit_gbo(&self, _submission: GboSubmission) -> Result<(), SdqError> {
        Err(SdqError::NotImplemented)
    }
}

File: sdq-analysis-api-java/domain/ports/incoming/service/src/main/java/uk/ratracejoe/sdq/service/GoalService.java
package uk.ratracejoe.sdq.service;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import uk.ratracejoe.sdq.model.Assessor;
import uk.ratracejoe.sdq.model.demographics.DemographicFilter;
import uk.ratracejoe.sdq.model.gbo.GboSubmission;
import uk.ratracejoe.sdq.model.gbo.Goal;
import uk.ratracejoe.sdq.model.gbo.GoalProgress;
import uk.ratracejoe.sdq.model.gbo.GoalType;

public interface GoalService {

  Goal createGoal(Goal goal);

  void recordGoalScore(GboSubmission gbo);

  List<Goal> getGoalsForClient(UUID clientId);

  List<GoalProgress> getGoalsWithProgress(
      List<Assessor> assessors,
      List<DemographicFilter> filters,
      int minProgress,
      List<GoalType> goalTypes,
      LocalDate from,
      LocalDate to);

  Goal updateGoal(Goal goal);

  Goal getGoal(UUID goalId);

  GoalProgress getGoalProgress(UUID goalId, Assessor assessor);

  List<GoalProgress> getGoalsProgressForClient(UUID clientId, Assessor assessor);
}
